### PR TITLE
berkeley-db: up formula version to 6.2.23, enable dbm, depend on java

### PR DESCRIPTION
**Pull request by maxim-belkin**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

This PR does several things:
1. updates `berkeley-db` to 6.2.23.
1. updates default options to build with java and sql support
1. adds `:java` dependency
1. adds `dbm` support (needed for some formulas)

Related (and closed) PRs: #2549 and #3263

```txt
$ brew audit --strict berkeley-db
berkeley-db:
  * JARs were installed to "/usr/local/Cellar/berkeley-db/6.2.23/lib"
    Installing JARs to "lib" can cause conflicts between packages.
    For Java software, it is typically better for the formula to
    install to "libexec" and then symlink or wrap binaries into "bin".
    See "activemq", "jruby", etc. for examples.
    The offending files are:
      /usr/local/Cellar/berkeley-db/6.2.23/lib/db.jar
Error: 1 problem in 1 formula
```

### CODE REVIEW

File: Formula/berkeley-db.rb
Comment (MikeMcQuaid) [inline Formula/berkeley-db.rb]: Shouldn't build with Java by default because macOS doesn't ship with Java by default.

**Review by MikeMcQuaid — CHANGES_REQUESTED** [review]

**Comment by maxim-belkin** [comment]
Looks like there is a number of failures due to non-existent 
```
/usr/local/opt/berkeley-db/lib/libdb-6.1.dylib
```

**Comment by zmwangx** [comment]
I attempted this in #2549. rpm was the only blocker back then.

**Comment by maxim-belkin** [comment]
@zmwangx yes, I saw that one and referenced it in the original post

**Comment by zmwangx** [comment]
Ah right. Just wanted to point it out to potentially save some research.

**Comment by MikeMcQuaid** [comment]
@BrewTestBot test this please

**Comment by maxim-belkin** [comment]
What's going on with downloaded files? They have the same SHA256 :)
```
=> brew fetch --retry bogofilter --build-bottle --force
==> FAILED
==> Downloading https://downloads.sourceforge.net/project/bogofilter/bogofilter-1.2.4/bogofilter-1.2.4.tar.bz2
==> Downloading from http://downloads.sourceforge.net/#!/project/bogofilter/bogofilter-1.2.4/bogofilter-1.2.4.tar.bz2
Downloaded to: /Users/brew/Library/Caches/Homebrew/bogofilter-1.2.4.tar.bz2
SHA256: 818a4b8bbcb50878a8b1b9f71b4274d242ab46bf860c74676e98dec1d0248821
```
and
```
==> brew fetch --retry darkice --build-bottle --force
==> FAILED
==> Downloading https://downloads.sourceforge.net/project/darkice/darkice/1.3/darkice-1.3.tar.gz
==> Downloading from http://downloads.sourceforge.net/#!/project/darkice/darkice/1.3/darkice-1.3.tar.gz
Downloaded to: /Users/brew/Library/Caches/Homebrew/darkice-1.3.tar.gz
SHA256: 818a4b8bbcb50878a8b1b9f71b4274d242ab46bf860c74676e98dec1d0248821
```

**Comment by maxim-belkin** [comment]
And thanks for starting this test because it is blocking some of my work in Linuxbrew-land.

**Comment by maxim-belkin** [comment]
@MikeMcQuaid could you please restart the build?

**Comment by maxim-belkin** [comment]
my gist-logs for [rpm](https://gist.github.com/1d688342a9349ee589f44cf98f34c3fd) and [bogofilter](https://gist.github.com/753a60127e5210295f31afb9175b85fd)

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by maxim-belkin** [comment]
fixed `rpm` test locally.
I have a solution for `darkice` but it requires modifications to several source code files in `darkice` OR one header file in `libsamplerate`

**Comment by maxim-belkin** [comment]
Fixed `darkice` and `rpm`'s testing section. Ready to be re-tested ( @MikeMcQuaid ) .

**Comment by maxim-belkin** [comment]
Weird failure on El Capitan: the only part that failed is the development build of `moc`. Note that El Capitan uses SDK for `10.12`

```log
==> brew install --devel moc --verbose
==> FAILED
```
then:
```log
==> Installing moc
==> Using the sandbox
/usr/bin/sandbox-exec -f /tmp/homebrew20161209-86963-1nsoq7g.sb nice /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/bin/ruby -W0 -I /usr/local/Homebrew/Library/Homebrew -- /usr/local/Homebrew/Library/Homebrew/build.rb /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/moc.rb --verbose --devel
==> Downloading http://ftp.daper.net/pub/soft/moc/unstable/moc-2.6-alpha2.tar.xz
Already downloaded: /Users/brew/Library/Caches/Homebrew/moc-2.6-alpha2.tar.xz
==> Verifying moc-2.6-alpha2.tar.xz checksum
tar xJf /Users/brew/Library/Caches/Homebrew/moc-2.6-alpha2.tar.xz
==> ./configure --disable-debug --prefix=/usr/local/Cellar/moc/2.6-alpha2_1
...
... some output ...
...
Making install in vorbis
/bin/sh ../../libtool  --tag=CC   --mode=compile clang -DHAVE_CONFIG_H -I. -I../..    -I/usr/local/Cellar/libogg/1.3.2/include -I/usr/local/Cellar/libvorbis/1.3.5/include -I../.. -D_THREAD_SAFE -pthread -g -O2 -Wall -W -I/usr/local/Cellar/jack/0.125.0_1/include -I/usr/local/Cellar/libsamplerate/0.1.9/include  -MT libvorbis_decoder_la-vorbis.lo -MD -MP -MF .deps/libvorbis_decoder_la-vorbis.Tpo -c -o libvorbis_decoder_la-vorbis.lo `test -f 'vorbis.c' || echo './'`vorbis.c
libtool: compile:  clang -DHAVE_CONFIG_H -I. -I../.. -I/usr/local/Cellar/libogg/1.3.2/include -I/usr/local/Cellar/libvorbis/1.3.5/include -I../.. -D_THREAD_SAFE -pthread -g -O2 -Wall -W -I/usr/local/Cellar/jack/0.125.0_1/include -I/usr/local/Cellar/libsamplerate/0.1.9/include -MT libvorbis_decoder_la-vorbis.lo -MD -MP -MF .deps/libvorbis_decoder_la-vorbis.Tpo -c vorbis.c  -fno-common -DPIC -o .libs/libvorbis_decoder_la-vorbis.o
In file included from vorbis.c:35:
In file included from ../../decoder.h:6:
In file included from ../../io.h:5:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk/usr/include/pthread.h:60:
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk/usr/include/time.h:177:5: error: conflicting types for 'clock_gettime'
int clock_gettime(clockid_t __clock_id, struct timespec *__tp);
    ^
../../compat.h:77:5: note: previous declaration is here
int clock_gettime (int clk_id, struct timespec *ts);
    ^
1 error generated.
make[2]: *** [libvorbis_decoder_la-vorbis.lo] Error 1
make[1]: *** [install-recursive] Error 1
make: *** [install-recursive] Error 1

==> Formula
Path: /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/moc.rb
```

**Comment by ilovezfs** [comment]
This fixes it
```
diff --git a/Formula/moc.rb b/Formula/moc.rb
index 12c9aa7..9d39a8c 100644
--- a/Formula/moc.rb
+++ b/Formula/moc.rb
@@ -45,6 +45,11 @@ class Moc < Formula
   depends_on "homebrew/dupes/ncurses" => :optional
 
   def install
+    # Fix "error: conflicting types for 'clock_gettime'"
+    if build.devel? && MacOS.version == :el_capitan && MacOS::Xcode.installed? && MacOS::Xcode.version >= "8.0"
+      inreplace Dir["{*.c,compat.h}"], "clock_gettime", "moc_clock_gettime", false
+    end
+
     system "autoreconf", "-fvi" if build.head?
     system "./configure", "--disable-debug", "--prefix=#{prefix}"
     system "make", "install"
```
@maxim-belkin mind reporting it upstream to moc? :innocent:

**Comment by maxim-belkin** [comment]
A perfect stranger ( wink-wink @zmwangx ) already reported it [here](https://moc.daper.net/node/1576)
I can implement your fix and bump up the revision of development version of moc to 2.5.3. Should I do that?

**Comment by ilovezfs** [comment]
@maxim-belkin the name rings a bell.

>I can implement your fix and bump up the revision of development version of moc to 2.5.3. Should I do that?

Pretty please, with sugar on top. Also add a link to rings-a-bell's report.

**Comment by zmwangx** [comment]
I've been having some back-and-force with upstream maintainer John Fitzgerald (over private email), with me explaining the clock_gettime clusterfuck and both of us trying to come up ways to fix this in upstream code base satisfactorily. For instance, I've suggested `AC_CHECK_FUNCS` or `AC_SEARCH_LIBS` (reminding him of `SDKROOT` and our `ac_cv_func_clock_gettime` and `ac_cv_search_clock_gettime`); or renaming the function and define it unconditionally for Darwin:

```diff
diff --git a/compat.c b/compat.c
index df02a1d..5b0bcc5 100644
--- a/compat.c
+++ b/compat.c
@@ -49,13 +49,13 @@ char *strcasestr (const char *haystack, const char *needle)
#endif

/* OSX doesn't provide clock_gettime(3) so fall back to gettimeofday(2). */
-#ifndef HAVE_CLOCK_GETTIME
+#if !defined HAVE_CLOCK_GETTIME || defined __APPLE__
#include <sys/time.h>
#include <assert.h>

#include "common.h"

-int clock_gettime (int clk_id ASSERT_ONLY, struct timespec *ts)
+int my_clock_gettime (int clk_id ASSERT_ONLY, struct timespec *ts)
{
	int result;
	struct timeval tv;
diff --git a/compat.h b/compat.h
index 71a011e..27c0283 100644
--- a/compat.h
+++ b/compat.h
@@ -50,10 +50,12 @@ extern "C" {
char *strcasestr (const char *haystack, const char *needle);
#endif

-#ifndef HAVE_CLOCK_GETTIME
+#if !defined HAVE_CLOCK_GETTIME || defined __APPLE__
#define CLOCK_REALTIME 0
struct timespec;
-int clock_gettime (int clk_id, struct timespec *ts);
+int my_clock_gettime (int clk_id, struct timespec *ts);
+#else
+#define my_clock_gettime clock_gettime
#endif

#ifdef __cplusplus
diff --git a/interface.c b/interface.c
index e8c3e08..976f023 100644
--- a/interface.c
+++ b/interface.c
@@ -2581,7 +2581,7 @@ static time_t rounded_time ()
	struct timespec exact_time;
	time_t curr_time;

-	if (clock_gettime (CLOCK_REALTIME, &exact_time) == -1)
+	if (my_clock_gettime (CLOCK_REALTIME, &exact_time) == -1)
		interface_fatal ("clock_gettime() failed: %s", xstrerror (errno));

	curr_time = exact_time.tv_sec;
```

He was unsatisfied with both approaches and decided that he would use a `host_os` based check (no no to me) in his latest email to me (from yesterday).

Whatever the outcome is, rest assured that upstream really wants to fix this.

**Comment by ilovezfs** [comment]
@zmwangx you may want to share this with him: https://github.com/jsquyres/libfabric/commit/469779935f9995b88a6f486fe3c4c4496989af36

**Comment by zmwangx** [comment]
@ilovezfs We've already discussed `AC_LINK_IFELSE` and such, I'm sure he won't be thrilled to carry that `ofi_clock_gettime.m4` in moc's codebase. However, feel like contributing to Autoconf Archive, or asking Jeff Squyres to do that?

**Comment by ilovezfs** [comment]
@zmwangx considering Jeff actually decided to remove clock_gettime entirely from their codebase that somehow feels wrong lol

**Comment by ilovezfs** [comment]
@zmwangx https://github.com/ofiwg/libfabric/pull/2508.

**Comment by zmwangx** [comment]
@ilovezfs John considered that (there's only one clock_gettime call after all, and it's not even essential, it's only called when seeking forward/backward), but he had concerns over precision of alternate implementations.

I feel bad for all these poor Linux souls who are now forced to clean after Apple's mess.

**Comment by zmwangx** [comment]
By the way, re rpm @maxim-belkin: are you sure `/var/lib/rpm/sqldb` is the correct replacement for `/var/lib/rpm/Packages`? Last time I looked at the this I thought it might be due to a subtly broken build, but I don't know enough about rpm (well, I know almost nothing) to tell either way.

**Comment by ilovezfs** [comment]
@zmwangx there are two separate issues really: how to detect the evil case, and what to do about it. Assuming the former is not a problem, it seems like changing the name of the function, using a toll free bridge when the platform has the real clock_gettime, and otherwise using one of the preexisting compat implementations is pretty uncontroversial, no?

**Comment by zmwangx** [comment]
@ilovezfs 

> ... pretty uncontroversial, no?

To me it's pretty uncontroversial. However, I wrote

> 3. Ignore Darwin's clock_gettime. Something like the patch attached at
  the end.

and John wrote

> Yeah, I did come up with a similar patch, but decided it got too cryptic.

A little bit of context: a while ago John discarded a lot of compat stuff and tried to make the code base as POSIX as possible. It's understandable he doesn't want something like `moc_clock_gettime` in the C code. But then you into all the problems with the heavily contended symbol `clock_gettime`.

**Comment by ilovezfs** [comment]
Given the limitations of the language, a wrapper seems preferable to littering the code with #if this #else that macros.

**Comment by zmwangx** [comment]
@ilovezfs You will have cpp conditionals anyway. A point can be made to limit tweaks to `compat.{c,h}` and pretend elsewhere (i.e. `interface.c`) that you have unfettered access to realtime extensions.

**Comment by ilovezfs** [comment]
@zmwangx yup, which is why "It's understandable he doesn't want something like moc_clock_gettime in the C code," is not actually understandable to me.

**Comment by zmwangx** [comment]
@ilovezfs Well, if you could "pretend elsewhere (i.e. interface.c) that you have unfettered access to realtime extensions", then you would just call `clock_gettime`, not `moc_clock_gettime`...

**Comment by ilovezfs** [comment]
@zmwangx I see, but tough to avoid including time.h indirectly at some point, no?

**Comment by zmwangx** [comment]
@ilovezfs Never mind, got the patch: https://raw.githubusercontent.com/Homebrew/patches/b7814c0/moc/moc-r2935%2Bclock_gettime-3.patch

Basically:

```
dnl OSX / MacOS doesn't provide clock_gettime(3) prior to darwin-16.0.0
dnl so fall back to gettimeofday(2).
case "$host_os" in
[*darwin[1-9].*] | [*darwin1[0-5].*])
	HAVE_CLOCK_GETTIME="no"
	;;
*darwin*)
	HAVE_CLOCK_GETTIME="yes"
	;;
*)
	AC_CHECK_LIB([rt], [clock_gettime],
		[HAVE_CLOCK_GETTIME="yes"],
		[HAVE_CLOCK_GETTIME="no"])
	;;
esac
```

```diff
diff --git a/Formula/moc.rb b/Formula/moc.rb
index 12c9aa70cd..0756bae400 100644
--- a/Formula/moc.rb
+++ b/Formula/moc.rb
@@ -11,9 +11,18 @@ class Moc < Formula
   end
 
   devel do
-    url "http://ftp.daper.net/pub/soft/moc/unstable/moc-2.6-alpha2.tar.xz"
-    sha256 "0a3a4fb11227ec58025f7177a3212aca9c9955226a2983939e8db662af13434b"
+    url "http://ftp.daper.net/pub/soft/moc/unstable/moc-2.6-alpha3.tar.xz"
+    sha256 "a27b8888984cf8dbcd758584961529ddf48c237caa9b40b67423fbfbb88323b1"
 
+    patch do
+      url "https://raw.githubusercontent.com/Homebrew/patches/b7814c0/moc/moc-r2935%2Bclock_gettime-3.patch"
+      sha256 "601b5cdf59db67f180f1aaa6cc90804c1cb69c44cdecb2e8149338782e4f21a8"
+    end
+
+    # Remove build deps when next devel release comes out
+    depends_on "autoconf" => :build
+    depends_on "automake" => :build
+    depends_on "gettext" => :build
     depends_on "popt"
   end
 
@@ -45,7 +54,8 @@ class Moc < Formula
   depends_on "homebrew/dupes/ncurses" => :optional
 
   def install
-    system "autoreconf", "-fvi" if build.head?
+    # Remove build.devel? when next devel release comes out
+    system "autoreconf", "-fvi" if build.head? || build.devel?
     system "./configure", "--disable-debug", "--prefix=#{prefix}"
     system "make", "install"
   end
```

**Comment by MikeMcQuaid** [comment]
> +    # Remove build deps when next devel release comes out

Probably worth noting a specific version or even adding a `test` or this will get left indefinitely.

**Comment by zmwangx** [comment]
> Probably worth noting a specific version or even adding a test or this will get left indefinitely.

My patch is only meant for testing... I think

```rb
inreplace Dir["{*.c,compat.h}"], "clock_gettime", "moc_clock_gettime", false
```

is good enough for now.

**Comment by zmwangx** [comment]
By the way, the `clock_gettime` issue is subtle in moc. Even if `clock_gettime` is linked on 10.11, running `mocp` won't fail immediately; only if you actually play a song in it and try to seek forward or backward will it error out with the dreaded

```
dyld: lazy symbol binding failed: Symbol not found: _clock_gettime
```

Of course we can't test that.

The only reliable test that I know of is

```rb
system "/usr/bin/nm #{bin}/mocp | grep clock_gettime"
```

**Comment by maxim-belkin** [comment]
build on sierra failed...

**Comment by zmwangx** [comment]
> build on sierra failed...

Because the replacement needs to happen on 10.12 too (`HAVE_CLOCK_GETTIME` isn't set there thanks to looking for `clock_gettime` in the nonexistent librt). Now that there's an official patch though, let's just use that one (at the expense of having to regenerate configure).

I've opened #7734 to bump moc devel so that we don't need to keep rebuilding everything else here. Might as well rebase after #7734 is merged; berkeley-db 6.2.23 has been delayed for the better part of a year.

**Comment by maxim-belkin** [comment]
OK, I'll rebase this PR and remove all the fixes for `moc` once your PR is merged.

**Comment by maxim-belkin** [comment]
rebased and force-pushed

**Comment by maxim-belkin** [comment]
@ilovezfs have a look at error messages on Jenkins -- I think those errors are Jenkins' problem. Can you restart the build?

**Comment by maxim-belkin** [comment]
@zmwangx regarding `rpm`: I don't know but I see `sqldb`, not `Packages` when I run tests. I checked that this file is an sql database with `Packages` table in it. It could be a combination of factors that leads to `sqldb` file and not a bunch of Berkeley-DB files one of which if `Packages`

**Comment by ilovezfs** [comment]
@BrewTestBot test this please

**Comment by maxim-belkin** [comment]
Strange failures again. Is there something wrong with Jenkins or does it indicate a problem with formulae?

**Comment by MikeMcQuaid** [comment]
Will try to rebuild.

**Comment by MikeMcQuaid** [comment]
@BrewTestBot test this please

**Comment by maxim-belkin** [comment]
Can someone take a look at this PR? Looks like all the checks are passed now.

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by MikeMcQuaid** [comment]
Looks good to me. @ilovezfs 👍 ?

File: Formula/darkice.rb
Comment (ilovezfs) [inline Formula/darkice.rb]: Has this been reported?

**Comment by ilovezfs** [comment]
Does it really still need ENV.deparallelize for berkeley-db?

File: Formula/berkeley-db.rb
Comment (ilovezfs) [inline Formula/berkeley-db.rb]: if sql and dbm don't add dependencies, why are they optional as opposed to always passed?

File: Formula/berkeley-db.rb
Comment (ilovezfs) [inline Formula/berkeley-db.rb]: is it really only a build time dep?

File: Formula/darkice.rb
Comment (maxim-belkin) [inline Formula/darkice.rb]: Thanks for asking. I was going to say "no", but then looked around and found [this PR](https://github.com/rafael2k/darkice/pull/120). So yes, this has been reported

Comment (ilovezfs) [inline Formula/darkice.rb]: Cool. Please add the link, date, and brief description like these: https://gist.github.com/ilovezfs/62a9eb715dc044f7202cbf86caaf713b

Comment (maxim-belkin) [inline Formula/darkice.rb]: within the `echo do` loop or right before it?

File: Formula/berkeley-db.rb
Comment (maxim-belkin) [inline Formula/berkeley-db.rb]: Do you mean we should convert "without-sql" to "with-sql" OR `depends_on "sql" => :build` ? I guess the latter but want to make sure

**Comment by maxim-belkin** [comment]
I've taken all suggestions into account and now formula builds with `sql` and `dbm` by default:
```configure
--enable-sql
--enable-sql_codegen
--enable-dbm
```
I don't see any additional dependencies because of this change.

`java` is `build` only dependency
parallel build (still) fails.

[Options](http://docs.oracle.com/cd/E17076_04/html/installation/build_unix_conf.html) to consider:

`--enable-sql_compat`: builds sqlite3.

I see that `sqlite` formula is keg-only because of `sqlite3` and that's why I didn't enable this option in this PR. Let me know if I should add it here.

**Comment by maxim-belkin** [comment]
added `--enable-stl`

**Comment by maxim-belkin** [comment]
@ilovezfs I've addressed your comments/suggestions

**Review by ilovezfs — APPROVED** [review]

**Comment by ilovezfs** [comment]
@maxim-belkin thanks for your contribution to Homebrew! I did some minor surgery on the PR so that it's one commit per formula and so that it uses the standard commit subjects, but preserved your commit credits for the changes to berkeleydb, darkice, and rpm.

**Comment by maxim-belkin** [comment]
Cool! Thank you!

**Pull request closed without merging**
